### ISSUE TITLE
Delayed one-time execution for Jobs

### ISSUE THREAD

**Issue by vaizki** [kubernetes/kubernetes#75817]
**What would you like to be added**:
- a new Time property in the JobSpec to indicate a time before which the Job should not be scheduled (name TBD)
- changes to Job controller to defer sync/scheduling of Jobs with the new property until such time

**Why is this needed**:
To complement CronJobs with a one-shot scheduled Job, similar to *NIX `at` jobs. 

The alternatives are a bit heavy:
- create a custom controller (for example metacontroller CompositeController) with a CRD AtJob
- deploy Airflow, Rundeck or other blob of additional stack

**Proposal**:
Having had a look at the controller implementation it seems like a minor change with the following contract:
- Jobs are guaranteed not to be scheduled before the specified time 
- There is no guarantee how soon they will run after the specified time but the granularity should be documented
- Same completion guarantees as Jobs now have
- Same Job cleanup, ttl etc mechanisms would apply

Property name could be for example `deferUntil:` or `executeAfter:`

/sig apps

**Comment by danielqsj** [comment]
Does cronjob not enough in this case ?
For instance, does `45 23 * * *` means job will scheduler after 23:45 every day ?
If so, why need another filed to define when job should be scheduled ?

**Comment by vaizki** [comment]
@danielqsj CronJob indeed schedules for example every day after 23:43. The idea in this enhancement is to schedule the Job once for a certain singe point of time and when it has reached the desired completions then it will never be run again.

**Comment by draveness** [comment]
Maybe we could have something like the liveness probe in the pod which check the context before actually schedule the job.

```apiVersion: batch/v1
kind: Job
metadata: # metadata for the job
  name: pi-job
spec:
  template: # template for the pod the job will create.
    metadata: # metadata for the pod
      name: pi-pod
    spec:
      containers:
      - name: pi
        image: perl
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  schedulesIf:
      exec:
        command:
        - cat
        - /tmp/healthy
```

**Comment by vaizki** [comment]
> Maybe we could have something like the liveness probe in the pod which check the context before actually schedule the job.

What would these probes run against if the Pod has not been scheduled yet?

**Comment by pentago** [comment]
Let's continue..

**Comment by piyushsingariya** [comment]
This issue should be considered. There are many use-cases for such feature.

**Comment by to266** [comment]
Just FYI, the immediate alternative to me seems to be:
 - Schedule a cronjob with the exact date for the wished job (i.e. so that it runs once a year)
 - Remove the cronjob at some point after the scheduled date, but not later than 1 year after that.
   - Can be further (over) complicated by creating another similarly-scheduled cronjob, but it would remove first the original JOB, and then itself

**Comment by sremp** [comment]
Kudos to the custom operator solution. I had a similar need previously, the workaround I developed involved a lot of finicky wiring up Jenkins pipelines and Ansible playbooks with k8s. It got the job done but was not a great solution.


For anyone looking delaying the execution of a generic Job after creation: 
I had a use case where I need to run a Deployment via Flux and then execute a post-deploy Job a short time after, but _not_ before the API is available.  What I ended up doing was to add a `sleep` command to wait 10 seconds before running the bootstrap script: 

```
command: ["/bin/sh", "-c"]
args:
  - sleep 5;
     ./foo
```
